server/routerlicious/packages/services-utils/src/auth.ts

```diff
@@ -96,29 +89,37 @@ export function verifyStorageToken(tenantManager: ITenantManager, config: Provid
         const isTokenExpiryEnabled = config.get("auth:enableTokenExpiration") as boolean;
         const authorizationHeader = request.header("Authorization");
         if (!authorizationHeader) {
-            return res.status(401).send("Authorization header is missing.");
+            return res.status(403).send("Missing Authorization header.");
         }
         const regex = /Basic (.+)/;
         const tokenMatch = regex.exec(authorizationHeader);
         if (!tokenMatch || !tokenMatch[1]) {
-            return res.status(401).send("Missing access token.");
+            return res.status(403).send("Missing access token.");
         }
         const token = tokenMatch[1];
         const tenantId = getParam(request.params, "tenantId");
         const documentId = getParam(request.params, "id") || request.body.id;
         if (!tenantId || !documentId) {
-            return res.status(401).send("TenantId or DocumentId is missing in the access token.");
+            return res.status(403).send("Missing tenantId or documentId in request.");
         }
-        const claims = validateTokenClaims(token, documentId, tenantId, maxTokenLifetimeSec, isTokenExpiryEnabled);
-        if (!claims) {
-            return res.status(401).send("Invalid access token.");
+        let claims: ITokenClaims;
+        try {
+            claims = validateTokenClaims(token, documentId, tenantId);
+            if (isTokenExpiryEnabled) {
+                validateTokenClaimsExpiration(claims, maxTokenLifetimeSec);
+            }
+        } catch (error) {
+            if (error instanceof NetworkError) {
+                return res.status(error.code).send(error.message);
+            }
+            throw error;
         }
         tenantManager.verifyToken(claims.tenantId, token)
             .then(() => {
                 next();
             })
             .catch((error) => {
-                return res.status(401).json(error);
+                return res.status(403).json(error);
```

Comment:
switching from 401 to 403 will require a change in the Azure Fluid documentation as well. That is the documentation that customers refer to, while onboarding to FRS.

Curious to know why we are not returning 401? I understand that 401 is being considered as retriable at driver side. But may be we should 
1.  make change at driver side to not retry for 401 and ,
2. on server side to send 500 (or something apt) for retry.

Reply:
Oh! it's about making the documentation change. That should be easy.

I just wanted to understand the reasoning behind this change. Also, will this be consistent among all the services, including historian? I can make the documentation change once we can confirm that for token validation failure, services will always send 403.

Reply:
Yes the plan is to have the same return semantics in historian as well

Reply:
> I can make the documentation change once we can confirm that for token validation failure, services will always send 403.
> — @sharmakhushboo

> It's possible to use a combination of 401 and 403 but this will be more confusing to reason about.
> — @tanviraumi 

My thinking was that we should use [401 Unauthorized](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401) when a token is invalid but authentication is possible (i.e. token is expired, client needs to refresh the token it is using), and we should use [403 Forbidden](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403) when a token is invalid but authorization is not possible (i.e. incorrect tenant key, mismatched tenant/document ids in request, insufficient scopes for action, etc.).

Without using a combination of 401 and 403, it will be difficult to determine when to bother retrying an auth error with a refreshed token. Is it possible to say, in the documentation, that services will return a 403 when a client is unauthorized and 401 when a client is unauthenticated?

Reply:
I think we should be more explicit and say we don't retry for 403 so you should get a new valid token from host. For 401, we will retry internally and it should work behind the scene.

Reply:
Yes, we could add that distinction but wouldn't it be better to just have one HttpStatusCode response with an error code that explains the error. The errorCode could be standardized and published in the documentation as well.

for example: 
403, errorCode: TokenExpired, errorMsg: <friendly msg>
403, errorCode: InvalidToken, errorMsg: <friendly msg>

In both cases, the error is non-retriable. Based on the errorCode, C1 decided whether to refresh or fix the token.

P.S. I hope you meant unauthorized for 401, as we don't authenticate C1. We only authorize. :p

Reply:
I meant that 401 would mean "you're probably unauthenticated, go get authenticated then we'll tell you whether you're authorized."

> In both cases, the error is non-retriable. Based on the errorCode, C1 decided whether to refresh or fix the token.

The point of a 401 (expired token) is that it _is_ retriable, and we can do the necessary logic behind the scenes in the driver to do that retry by fetching a refreshed token using the TokenProvider provided by C1. I'm not sure why we would add a "standardized" alternative error code to the already existing standardized error codes 401/403 :/

Reply:
We could be specific in the documentation, for example:

> 401: token is expired
> 403: token is invalid, check error message for more details

Reply:
okay. that makes sense. If the driver code is going to retry on 401, which like you said is not handled by the client/C1, it works.
So for C1, service will always send 403. Is that correct?

Reply:
I think we are getting confused between C1 and C2. C1s job is to implement the TokenProvider. All error codes are directly going to C2, who is running the driver code. If it's a 401, driver auto retries behind the scene and tries to fix the error. If its a 403, driver gives up and C2 will face a non-recoverable error. At this point, C2 will usually refresh the browser and reestablish the communication with C1, which may or may not fix the problem. This contract is between C1 and C2 (via the host) and we don't control that. We never send anything to C1 directly either

Reply:
Yes, that's correct.
As per my understand, from Azure Fluid documentation perspective, 403 is what will be important. As 401 would be taken care internally by the driver. How 403 is handled, would depend on how C1's application surfaces the error to C2.

Reply:
Not sure how up do date the doc is (https://github.com/microsoft/FluidFramework/tree/main/packages/loader/container-loader#error-handling) and @znewton  can correct me. But I believe 403 will bubble up to the container, and then its upto application developer on how they want to handle it. We should definitely provide instructions on how to catch the error from Host POV

Reply:
Correct! Both a 403 and a 401 will bubble up to the container as an error with `errorType` property set to `DriverErrorType.authorizationError`, and `statusCode` set to `401` or `403`.

A 401 could be thrown by the driver in the event that C1 made a mistake when setting up their TokenProvider or Host, such as always setting exp to 0, or something like that, where it will always be expired and getting a new token will not resolve the 401. In that specific scenario, the Driver will bubble up a 401 to the container. In this case, C1's app would never work at all, so in theory C1 would catch this bug in development, and C2 would never run this broken code.

Regardless of whether it's a 401 or 403, the container will receive a `DriverErrorType.authorizationError` from the Driver, so the host will be able to handle it as such by listening to the "closed" event on the container as described in the doc @tanviraumi linked.

Reply:
Alright. Thanks for the explanation. We should update the documents to reflect this. @znewton I'll ping you about the documentation details.
Approving this pr. :)